lib/puppet/file_serving/terminus_helper.rb

```diff
@@ -29,4 +29,16 @@ def path2instances(request, *paths)
       path2instance(request, base_path, :ignore_source_permissions => true, :relative_path => file)
     end
   end
+
+  def self.unescape_url(url)
+    url.sub(/^url=/,'')
+  end
+
+  def self.escape_url(url)
+    if url =~ /^https?:/
+      "url=#{url}"
+    else
+      url
+    end
```

Comment:
Ack. So here's the full story:

    $ bundle exec puppet apply -e 'file { "/tmp/httptest": source => "http://cdn.sstatic.net/stackoverflow/img/newsletter-ad.png" }' --trace
    Notice: Compiled catalog for geras.fritz.box in environment production in 1.28 seconds
    Error: /Stage[main]/Main/File[/tmp/httptest]: Could not evaluate: Could not retrieve file metadata for http://cdn.sstatic.net/stackoverflow/img/newsletter-ad.png: Could not find a directory environment named 'stackoverflow' anywhere in the path: /home/ffrank/.puppetlabs/etc/code/environments. Does the directory exist?

This is the top of the stack:

    /home/ffrank/git/puppet/lib/puppet/environments.rb:38:in `get!'
    /home/ffrank/git/puppet/lib/puppet/indirector/request.rb:40:in `environment='
    /home/ffrank/git/puppet/lib/puppet/indirector/request.rb:255:in `set_uri_key'
    /home/ffrank/git/puppet/lib/puppet/indirector/request.rb:81:in `initialize'

The indirector tries to analyze the URI [here](https://github.com/puppetlabs/puppet/blob/6df5fd9250affd62a2d18fc2cfb8bafb5a144f92/lib/puppet/indirector/request.rb#L253). `file://` and `puppet://` URL do not suffer this issue because there is special handling for those just above that line.

I believe that the indirector API is too simple for us to do a clean implementation of "special" URIs that should not be treated this way (i.e., we have no way to tell the indirector that one request is different from another in this way).